### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by a4000**
Addresses https://github.com/nf-core/ampliseq/issues/612

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/ampliseq/tree/master/.github/CONTRIBUTING.md)
- [ ] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: modules/local/phyloseq.nf
Comment (erikrikarddaniel) [inline modules/local/phyloseq.nf:28]:
```suggestion
    def tree    = "\"${tree}\""
    def prefix  = "\"${prefix}\""
```

File: workflows/ampliseq.nf
Comment (erikrikarddaniel) [inline workflows/ampliseq.nf:463]:
```suggestion

```

File: workflows/ampliseq.nf
Comment (erikrikarddaniel) [inline workflows/ampliseq.nf:484]:
```suggestion
    }
```

File: workflows/ampliseq.nf
Comment (erikrikarddaniel) [inline workflows/ampliseq.nf:604]:
```suggestion
    }
```

**Review by erikrikarddaniel — APPROVED** [review]
Looks fine to me, but I know you discussed with @d4straub so maybe wait for another review from him.

**Comment by erikrikarddaniel** [comment]
Now I see some tests fail because of updates with 2.9 tools. @d4straub updated to the new template yesterday.

File: workflows/ampliseq.nf
Comment (d4straub) [inline workflows/ampliseq.nf:636]: Could you put this into a subworkflow? That would make it easier to maintain, I think?

File: modules/local/phyloseq.nf
Comment (d4straub) [inline modules/local/phyloseq.nf:57]: tiny formatting: I prefer to have that in multiple lines, easier to read imho.

Comment (a4000) [reply modules/local/phyloseq.nf:57]: Do you mean line 57? I agree that would make it easier to read.

**Review by d4straub — COMMENTED** [review]
That looks thorough, great!

Could you merge the recently added PR https://github.com/nf-core/ampliseq/pull/614? That should solve the failing test `nf-core linting / nf-core` . `nf-core linting / EditorConfig` is caused by trailing whitespaces, you could avoid that by using [VSC and nf-core extensions](https://nf-co.re/docs/contributing/editor_plugins#vscode), if you wished to (or just take care manually). `nf-core linting / PythonBlack` can be solved by [black](https://nf-co.re/docs/contributing/code_formatting#black).

File: modules/local/phyloseq.nf
Comment (d4straub) [reply modules/local/phyloseq.nf:57]: yes, line 57

File: tests/pipeline/iontorrent.nf.test
Comment (d4straub) [inline tests/pipeline/iontorrent.nf.test:42]: Great that you added it here, but the test (and others that produce phyloseq output) is failing because in `tests/pipeline/iontorrent.nf.test.snap` the line https://github.com/nf-core/ampliseq/blob/a70ff48db034dc1fde657d426e4724793b91992a/tests/pipeline/iontorrent.nf.test.snap#L16 has an additional  `PHYLOSEQ={R=4.3.0, phyloseq=1.44.0}, `. 
[I like those tests, but they do over complicate it a little with the versions in that line...]

Comment (a4000) [reply tests/pipeline/iontorrent.nf.test]: I removed those tests for now

File: workflows/ampliseq.nf
Comment (a4000) [reply workflows/ampliseq.nf:639]: I created a subworkflow and made a few small changes in other parts ampliseq.nf just to make it easier to add the subworkflow.

File: tests/pipeline/iontorrent.nf.test
Comment (d4straub) [reply tests/pipeline/iontorrent.nf.test]: Sorry for the confusing message by me. I meant not to remove the lines in `tests/pipeline/*.nf.test`. I meant the files `tests/pipeline/*.nf.test.snap` need to be updated as well. For example:
in the github test https://github.com/nf-core/ampliseq/actions/runs/5803773532/job/15746522221?pr=615 is in "Run nf-test"
```
Test Workflow main.nf

java.lang.RuntimeException: Different Snapshot: 
Found:
[
    "{ASSIGNSH={pandas=1.1.5, python=3.9.1}, BARRNAP={barrnap=0.9}, CUSTOM_DUMPSOFTWAREVERSIONS={python=3.11.0, yaml=6.0}, CUTADAPT_BASIC={cutadapt=3.4}, DADA2_DENOISING={R=4.1.1, dada2=1.22.0}, DADA2_FILTNTRIM={R=4.1.1, dada2=1.22.0}, DADA2_QUALITY1={R=4.1.1, ShortRead=1.52.0, dada2=1.22.0}, DADA2_TAXONOMY={R=4.1.1, dada2=1.22.0}, FASTQC={fastqc=0.11.9}, FORMAT_TAXRESULTS_STD={pandas=1.1.5, python=3.9.1}, ITSX_CUTASV={ITSx=1.1.3}, PHYLOSEQ={R=4.3.0, phyloseq=1.44.0}, RENAME_RAW_DATA_FILES={sed=4.7}, SBDIEXPORT={R=3.6.3}, VSEARCH_USEARCHGLOBAL={vsearch=2.21.1}, Workflow={nf-core/ampliseq=2.7.0dev}}"
]

Expected:
[
    "{ASSIGNSH={pandas=1.1.5, python=3.9.1}, BARRNAP={barrnap=0.9}, CUSTOM_DUMPSOFTWAREVERSIONS={python=3.11.0, yaml=6.0}, CUTADAPT_BASIC={cutadapt=3.4}, DADA2_DENOISING={R=4.1.1, dada2=1.22.0}, DADA2_FILTNTRIM={R=4.1.1, dada2=1.22.0}, DADA2_QUALITY1={R=4.1.1, ShortRead=1.52.0, dada2=1.22.0}, DADA2_TAXONOMY={R=4.1.1, dada2=1.22.0}, FASTQC={fastqc=0.11.9}, FORMAT_TAXRESULTS_STD={pandas=1.1.5, python=3.9.1}, ITSX_CUTASV={ITSx=1.1.3}, RENAME_RAW_DATA_FILES={sed=4.7}, SBDIEXPORT={R=3.6.3}, VSEARCH_USEARCHGLOBAL={vsearch=2.21.1}, Workflow={nf-core/ampliseq=2.7.0dev}}"
]
  Test [192be031] 'Paired-End Pacbio ITS' FAILED (666.742s)
```
To solve that, in file `pacbio_its.nf.test.snap` the string 
`"{ASSIGNSH={pandas=1.1.5, python=3.9.1}, BARRNAP={barrnap=0.9}, CUSTOM_DUMPSOFTWAREVERSIONS={python=3.11.0, yaml=6.0}, CUTADAPT_BASIC={cutadapt=3.4}, DADA2_DENOISING={R=4.1.1, dada2=1.22.0}, DADA2_FILTNTRIM={R=4.1.1, dada2=1.22.0}, DADA2_QUALITY1={R=4.1.1, ShortRead=1.52.0, dada2=1.22.0}, DADA2_TAXONOMY={R=4.1.1, dada2=1.22.0}, FASTQC={fastqc=0.11.9}, FORMAT_TAXRESULTS_STD={pandas=1.1.5, python=3.9.1}, ITSX_CUTASV={ITSx=1.1.3}, RENAME_RAW_DATA_FILES={sed=4.7}, SBDIEXPORT={R=3.6.3}, VSEARCH_USEARCHGLOBAL={vsearch=2.21.1}, Workflow={nf-core/ampliseq=2.7.0dev}}"` 
needs to be replaced by 
`"{ASSIGNSH={pandas=1.1.5, python=3.9.1}, BARRNAP={barrnap=0.9}, CUSTOM_DUMPSOFTWAREVERSIONS={python=3.11.0, yaml=6.0}, CUTADAPT_BASIC={cutadapt=3.4}, DADA2_DENOISING={R=4.1.1, dada2=1.22.0}, DADA2_FILTNTRIM={R=4.1.1, dada2=1.22.0}, DADA2_QUALITY1={R=4.1.1, ShortRead=1.52.0, dada2=1.22.0}, DADA2_TAXONOMY={R=4.1.1, dada2=1.22.0}, FASTQC={fastqc=0.11.9}, FORMAT_TAXRESULTS_STD={pandas=1.1.5, python=3.9.1}, ITSX_CUTASV={ITSx=1.1.3}, PHYLOSEQ={R=4.3.0, phyloseq=1.44.0}, RENAME_RAW_DATA_FILES={sed=4.7}, SBDIEXPORT={R=3.6.3}, VSEARCH_USEARCHGLOBAL={vsearch=2.21.1}, Workflow={nf-core/ampliseq=2.7.0dev}}"`
note the additional `PHYLOSEQ={R=4.3.0, phyloseq=1.44.0}, `.

Comment (a4000) [reply tests/pipeline/iontorrent.nf.test:42]: Ah, I understand now. I've updated those .snap files now.

File: modules/local/phyloseq.nf
Comment (d4straub) [inline modules/local/phyloseq.nf:43]: I am not aware of any example for conditional input that uses `file.exists($file)`, so I am a little skeptic here. It might be working fine locally and on github, but we will need to see whether that works also on other systems. But easy enough to change in case it needs to be modified.

File: subworkflows/local/phyloseq_workflow.nf
Comment (d4straub) [inline subworkflows/local/phyloseq_workflow.nf:18]: If thats not a value then the process PHYLOSEQ will run only once, correct?
An alternative might be to combine all that channels, that is how its done in other parts of the code I think, for example [here](https://github.com/nf-core/ampliseq/blob/a70ff48db034dc1fde657d426e4724793b91992a/subworkflows/local/qiime2_ancom.nf#L19-L33).
Will the value channel still cause the file to be staged properly? If not, on HPC or cloud this method might fail while it succeeds locally and on github anyway.
Also that is something that could be fixed later if it appears to be indeed a problem.

**Review by d4straub — APPROVED** [review]
Looks good to me. Do you want to merge it?

File: modules/local/phyloseq.nf
Comment (a4000) [reply modules/local/phyloseq.nf:43]: That's a good point. I did my testing on a Linux server, but I'm hoping it would work on other systems.

File: subworkflows/local/phyloseq_workflow.nf
Comment (a4000) [reply subworkflows/local/phyloseq_workflow.nf:18]: That's correct, in my testing the process only ran once normally, but using first() to make it a value channel allowed the process to run multiple times for cases when there is multiple tax tables.

Comment (d4straub) [reply subworkflows/local/phyloseq_workflow.nf:18]: I checked with 
```
nextflow pull a4000/ampliseq
nextflow pull a4000/ampliseq -r dev
nextflow run a4000/ampliseq -r dev -profile test,singularity --skip_qiime --sbdiexport false --outdir results_a4000_dev
```
and the work directory had soft links to 
```
ASV_table.len.tsv
ASV_tax_species.gtdb.tsv
Metadata.tsv
```
so it seems despite the first() all is properly staged. Should be fine!

**Pull request merged**
